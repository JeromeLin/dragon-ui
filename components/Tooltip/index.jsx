import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

class Tooltip extends Component {
  render() {
    const { props } = this;
    const {
      title,
      direction,
      className,
      children,
      tipStyle,
      style,
    } = props;

    const cls = classnames({
      'ui-tooltip': true,
      [`ui-tooltip-${direction}`]: !!direction,
      [className]: !!className
    });
    return (
      <div className={cls} style={style}>
        <div className="ui-tooltip-inner" style={tipStyle}>
          {title}
        </div>
        {children}
      </div>
    );
  }
}

Tooltip.propTypes = {
  direction: PropTypes.oneOf([
    'top',
    'left',
    'right',
    'bottom',
    'topLeft',
    'topRight',
    'bottomLeft',
    'bottomRight',
    'leftTop',
    'leftBottom',
    'rightTop',
    'rightBottom'
  ])
};

Tooltip.defaultProps = {
  direction: 'bottom'
};

export default Tooltip;
