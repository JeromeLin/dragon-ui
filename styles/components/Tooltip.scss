%ui-tooltip-direction-left{
  left: -8px;
  &:before{
    top: 50%;
    right: -5px;
    transform: translateY(-50%);
    border-width: 5px 0 5px 5px;
    border-left-color: #000;
  }
}
%ui-tooltip-direction-right{
  right: -8px;
  &:before{
    top: 50%;
    left: -5px;
    transform: translateY(-50%);
    border-width: 5px 5px 5px 0;
    border-right-color: #000;
  }
}
%ui-tooltip-direction-top{
  top: -8px;
  &:before{
    bottom: -5px;
    left: 50%;
    margin-left: -5px;
    border-width: 5px 5px 0;
    border-top-color: #000;
  }
}
%ui-tooltip-direction-bottom{
  bottom: -8px;
  &:before{
    top: -5px;
    left: 50%;
    border-width: 0 5px 5px;
    margin-left: -5px;
    border-bottom-color: #000;
  }
}
.ui-tooltip{
  display: inline-block;
  position: relative;
  &:hover{
    .ui-tooltip-inner{
      display: block;
    }
  }
  .ui-tooltip-inner{
    display: none;
    max-width: 200px;
    padding: 3px 8px;
    color: #fff;
    text-align: center;
    background-color: #000;
    border-radius: 4px;
    position: absolute;
    z-index:1;
    &:before{
      content: '';
      position: absolute;
      width: 0;
      height: 0;
      border-color: transparent;
      border-style: solid;
    }
  }
  &.ui-tooltip-top{
     .ui-tooltip-inner{
       @extend %ui-tooltip-direction-top;
       left:50%;
       transform: translate3d(-50%, -100%, 0);
     }
   }
  &.ui-tooltip-topLeft{
     .ui-tooltip-inner{
       @extend %ui-tooltip-direction-top;
       left:0;
       transform: translate3d(-50%, -100%, 0);
     }
   }
  &.ui-tooltip-topRight{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-top;
      left:100%;
      transform: translate3d(-50%, -100%, 0);
    }
  }


  &.ui-tooltip-right{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-right;
      top:50%;
      transform: translate3d(100%, -50%, 0);
    }
  }
  &.ui-tooltip-rightTop{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-right;
      top:0;
      transform: translate3d(100%, -50%, 0);
    }
  }
  &.ui-tooltip-rightBottom{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-right;
      top:100%;
      transform: translate3d(100%, -50%, 0);
    }
  }

  &.ui-tooltip-bottom {
    .ui-tooltip-inner {
      @extend %ui-tooltip-direction-bottom;
      left: 50%;
      transform: translate3d(-50%, 100%, 0);
    }
  }
  &.ui-tooltip-bottomLeft{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-bottom;
      left:0;
      transform: translate3d(-50%, 100%, 0);
    }
  }
  &.ui-tooltip-bottomRight{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-bottom;
      left:100%;
      transform: translate3d(-50%, 100%, 0);
    }
  }


  &.ui-tooltip-left{
     .ui-tooltip-inner{
       @extend %ui-tooltip-direction-left;
       top:50%;
       transform: translate3d(-100%, -50%, 0);
     }
   }
  &.ui-tooltip-leftBottom{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-left;
      top:100%;
      transform: translate3d(-100%, -50%, 0);
    }
  }
  &.ui-tooltip-leftTop{
    .ui-tooltip-inner{
      @extend %ui-tooltip-direction-left;
      top:0;
      transform: translate3d(-100%, -50%, 0);
    }
  }
}