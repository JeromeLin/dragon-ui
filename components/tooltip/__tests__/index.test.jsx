import React from 'react';
import { render, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Tooltip from '../index';

if (global.document) {
  document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
  });
}

describe('Tooltip', () => {
  it('renders normal Tooltip correctly', () => {
    const wrapper = render(
      <Tooltip content="我是Tooltip内容">
        <div id="hello">Hello world!</div>
      </Tooltip>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('check `onVisibleChange` prop', () => {
    const onVisibleChange = jest.fn();

    const wrapper = mount(
      <Tooltip content="hello" trigger="click" onVisibleChange={onVisibleChange}>
        <div id="hello">Hello world!</div>
      </Tooltip>,
    );
    const div = wrapper.find('#hello');
    div.simulate('click');
    setTimeout(() => {
      expect(onVisibleChange).toHaveBeenLastCalledWith(true);
    });

    div.simulate('click');
    setTimeout(() => {
      expect(onVisibleChange).toHaveBeenLastCalledWith(false);
    });
  });

  it('check `visible` prop', () => {
    const onVisibleChange = jest.fn();
    const wrapper = mount(
      <Tooltip content="hello" trigger="manual" onVisibleChange={onVisibleChange}>
        <div id="hello">Hello world!</div>
      </Tooltip>,
    );
    wrapper.setProps({ visible: true });
    setTimeout(() => {
      expect(onVisibleChange).toHaveBeenLastCalledWith(true);
    });
  });
});
