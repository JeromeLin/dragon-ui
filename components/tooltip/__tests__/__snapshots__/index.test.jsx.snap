// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip renders normal Tooltip correctly 1`] = `
<div
  class="zw-tooltip"
>
  <div
    class="zw-tooltip__inner"
  >
    <div
      id="hello"
    >
      Hello world!
    </div>
  </div>
</div>
`;
